chapters/functions.tex

```diff
@@ -1880,10 +1880,11 @@ \section{Function Inlining and Event Generation}\label{function-inlining-and-eve
 "GenerateEvents" "=" ( false | true )
 \end{lstlisting}\end{synopsis}
 \begin{semantics}
-Has only an effect within a function declaration
-
-If \lstinline!GenerateEvents = true!, the model developer proposes that crossing functions in the function should generate events (one possibility of doing this is to inline the function and generate events for the inlined function).
+Has only an effect within a function declaration.
 
+If \lstinline!GenerateEvents = true!, the model developer proposes that crossing functions in the function should generate events.
+As defaults this implies \lstinline!Inline = true!, but it can be overriden by specifying one of the inline-annotations (with true value).
```

Comment:
I don't see clear support for allowing `LateInline` in the discussion yet, so either we need to also work on the definition of `LateInline`, or not include it for now:
```suggestion
This implies \lstinline!Inline = true! unless overridden by \lstinline!InlineAfterIndexReduction = true! (in particular, \lstinline!GenerateEvents = true! cannot be combined with \lstinline!Inline = false!).
```

Also note spelling of _overridden_.

Reply:
Here is something else that we should consider. All of those annotations (various Inlines and GenerateEvents) are formulated as "the model developer proposes..." To me it implies, that a tool is not required to follow any of those suggestions. And for inlining it makes sense: the result of the evaluation should be the same. For events, it actually changes the result. So does it make sense to keep it as "model developer proposes"?

Reply:
I fully agree with @eshmoylova's comment, and I would change "proposes" to "requires". That's the whole point of the `GenerateEvents` annotation: the modeller intends that events are indeed generated.

Reply:
Let us please continue this discussion either at line 1885 or 1887.  The suggestion for line 1886 has nothing to do with _proposes_ vs _requires_.

Reply:
I still don't see the problem with implementing LateInline, but I think the other change of GenerateEvents make this redundant, as the basic requirement is to generate events. So if a tool cannot generate events if inlining that late the alternative is to inline earlier.

Reply:
I agree that the situation is better now that `GenerateEvents` is imperative (still needs decision) while the inlining annotations are merely propositions.  Still, the combination with `LateInline` will appear problematic in tools where this corresponds to an inlining stage when it is too late to process `GenerateEvents = true`.  A non-normative remark could clarify two things:
- That tools should override any given inlining annotation that it can't respect in combination with `GenerateEvents = true`, instead performing the inlining at the nearest stage where `GenerateEvents = true` can still be respected.
- That a diagnostic is recommended in case inlining isn't performed at the stage proposed by the inlining annotation(s).

Reply:
I just realize that this happens to be the place where the non-normative text should go, so here's a (un-tested) suggestion:
```suggestion
This implies \lstinline!Inline = true! unless overridden by specifying one of the inlining annotations with value \lstinline!true! (in particular, \lstinline!GenerateEvents = true! cannot be combined with \lstinline!Inline = false!).

\begin{nonnormative}
In case a given inlining annotation proposes to inline at a stage when it is too late for a tool to process \lstinline!GenerateEvents = true!, the tool should instead perform inlining of the function at the nearest stage where \lstinline!GenerateEvents = true! can still be respected.
In this case, a diagnostic is recommended.
\end{nonnormative}
```

Reply:
> * That tools should override any given inlining annotation that it can't respect in combination with `GenerateEvents = true`, instead performing the inlining at the nearest stage where `GenerateEvents = true` can still be respected.

I like this suggestion because it accounts for the different tools having different abilities without making one or the other illegal. Is it worth mentioning that GenerateEvents =true may result in different results in different tools or is it obvious?

Also the suggested change above has a typo: "instead instead".

Reply:
> Also the suggested change above has a typo: "instead instead".

Thanks, fixed.

Reply:
> Is it worth mentioning that GenerateEvents =true may result in different results in different tools or is it obvious?

I don't find it obvious, especially considering the effort we've made to avoid ambiguity due to uncertain application of inverse functions.  However, I suppose we could construct an example where `GenerateEvents = true` could make index reduction fail by inlining before index reduction and thereby making a `derivative` unavailable?

Reply:
It looks good. Except I will remove 'too late' as there could be tools having problem with doing it too early.